src/irmin/type.ml

```diff
@@ -532,23 +532,23 @@ end
 let equal = Equal.t
 
 module Compare = struct
-  let unit (_ : unit) (_ : unit) = 0
+  let unit (_ : unit) (_ : unit) = 0 [@@inline always]
 
-  let bool (x : bool) (y : bool) = compare x y
+  let bool (x : bool) (y : bool) = compare x y [@@inline always]
 
-  let char = Char.compare
+  let char x y = Char.compare x y [@@inline always]
 
-  let int (x : int) (y : int) = compare x y
+  let int (x : int) (y : int) = compare x y [@@inline always]
```

Comment:
On `4.08.0`, this code produces a:

```asm
        call    caml_int_compare@PLT
```

Which will call `return Val_long(COMPARE_INT(v1, v2))`/`return Val_long((intnat)(v1 > v2) - (intnat)(v1 - v2))`. If semantically, we just want an order, we can replace it by `let int a b = a - b`, and OCaml will produce:

```asm
        subq    %rsi, %rax
        incq    %rax
```

It's a micro-optimization and it will change the semantic of the `compare` function - so I'm not sure if we need to apply it or not.

Reply:
it's not a good idea to use `(-)` as a comparison function as it is not transitive, due to overflows:

```
# min_int - (-1) < 0;;
- : bool = true
# (-1) - 1 < 0;;
- : bool = true
# min_int - 1 < 0;;
- : bool = false
```

Reply:
Yes, you confirm my think :+1: